### DISCUSSION TITLE
Battery Temperature Sensor not populated in Home Assistant

### DISCUSSION THREAD

**Discussion by madmat17** [wlcrs/huawei_solar#613] [Ideas]
Hi Jörg and team,

I love this integration! Formerly I used node-red to pull data via modbus-TCP from my inverter (prior my HASSO times); this integration just awesome!
In the `sensor.py` file I found that it seems that pulling the battery temperature has been [prepared](https://github.com/heinemannj/huawei_solar/blob/42f17e6b69b3b73931a234ea4acdc80ca165871b/sensor.py#L90) but either the sensor isn't presented to Home Assistant, or something is broken in between.
I can see the **inverter's** internal temperature (after activating the sensor), but no sensor showing the **battery's** temperature (even though that one might be considered even more crucial).

Unfortunately my Python skills are terribly poor - otherwise I might have found the culprit and could've contributed to the code...

Best,
Mat

**Comment by Roving-Ronin** [comment]
@madmat17 

Checking the feature requests and there is this CLOSED request: https://github.com/wlcrs/huawei_solar/issues/541

If wanting to add sensors for batteries (need to cater for users with 2):

![image](https://github.com/wlcrs/huawei_solar/assets/108674933/affb8c80-fbdc-4971-bcd1-c7b2f3f07700)

Then see: https://github.com/wlcrs/huawei_solar/wiki/Defining-custom-sensors

FYI for Step 1:

**register_names.py** includes these two required names at:
Battery1 - https://gitlab.com/Emilv2/huawei-solar/-/blob/master/src/huawei_solar/register_names.py#L133
Battery2 - https://gitlab.com/Emilv2/huawei-solar/-/blob/master/src/huawei_solar/register_names.py#L174

**registers.py** includes these two entries at: 
Battery1 - https://gitlab.com/Emilv2/huawei-solar/-/blob/master/src/huawei_solar/registers.py?ref_type=heads#L737 
Battery2 - https://gitlab.com/Emilv2/huawei-solar/-/blob/master/src/huawei_solar/registers.py?ref_type=heads#L754

would like to see this added also, but I get lost in the final code parts myself. :-(

**Comment by madmat17** [comment]
Hm. That's weird.
I just double-checked the [register_names.py](https://gitlab.com/Emilv2/huawei-solar/-/blob/master/src/huawei_solar/register_names.py) and/or [registers.py](https://gitlab.com/Emilv2/huawei-solar/-/blob/master/src/huawei_solar/registers.py) and both seem to contain the battery temperature.

From registers.py:
`rn.STORAGE_UNIT_1_BATTERY_TEMPERATURE: I16Register("°C", 10, 37022, 1)`
-> matching my records and how I used to poll the register (register 37022, Int16, gain of 10, length of 1).

Don't get it...

**Reply by madmat17** [reply]
I found that the key `storage_unit_1_battery_temperature` was missing in the `sensors.py` as well in the language translation file (in my case the German one).
But even after adding it to those I don't see the sensor. I checked the debug log of the Huawai Solar integration and it seems that the register isn't being polled at all.

**Comment by madmat17** [comment]
Stupid me...
One should read more carefully. Everything needed can be found [here](https://github.com/wlcrs/huawei_solar/wiki/Defining-custom-sensors) and [here](https://github.com/wlcrs/huawei_solar/compare/c023168..user-defined-extra-sensors).
Works like charm, after adding the `EXTRA_SENSOR_DESCRIPTIONS` to the `sensor.py`.

**Reply by Roving-Ronin** [reply]
Care to share ?

**Reply by madmat17** [reply]
You simly need to add the green lines from the [diff-view](https://github.com/wlcrs/huawei_solar/compare/c023168..user-defined-extra-sensors) to your `sensor.py` file and restart HA (whereas I strongly recommend to create a backup of your `sensor.py` - just in case).

**Reply by spelzl** [reply]
Can this change possibly be integrated into the generic HA integration or does everyone need to do it individually?

**Reply by Roving-Ronin** [reply]
@spelzl the original request for adding battery sensors is https://github.com/wlcrs/huawei_solar/issues/74. It appears to be still open (but labelled 'won't fix') and appears to be in that state to allow people to post if they think these sensor(s) should be added.

Suggest everyone after this goes to that post and leaves a comment.

**Reply by spelzl** [reply]
Thank you for your advice and link to the initial request.
I have added a comment there.

**Reply by Roving-Ronin** [reply]
Tried this and got it working with a dual LUNA installation.

Unfortunately though it ends up using the old sensor entity id naming format, so they end up being like this with battery repeated twice:

![image](https://github.com/wlcrs/huawei_solar/assets/108674933/e80b8b9e-1976-459c-99f0-184f8322b1b1)

Would anyhow have any idea of how to fix this AND correct them to the new format that would be:

sensor.battery_temperature (for #1) and sensor.battery_temperature_2 (for #2 and subsequent)?

**Comment by Roving-Ronin** [comment]
When you update WLCRS it will overwrite the custom version of sensor.py with the support for battery temperature added.
If anyone wants to avoid the headache of trying to work out how to do this with v.1.3.2 (the default file doesn't even include the section that says ' add extra sensors here' anymore) then you can download a version of v1.3.2 that has been modified to add these temperature sensors (for battery 1 and 2) back in from:

https://github.com/Roving-Ronin/huawei_solar
